src/main/java/com/intuit/graphql/orchestrator/batch/QueryOptimizer.java

```diff
@@ -0,0 +1,115 @@
+                                    .map(s -> (Field) s)
+                                    .collect(Collectors.toList())));
+                } else selectionSetMap.put(f.getName(), new HashSet<>(
+                        f.getSelectionSet().getSelections()
+                                .stream()
+                                .map(s -> (Field) s)
+                                .collect(Collectors.toList())));
+                createSelectionSetTree(f.getSelectionSet(), selectionSetMap);
+            }
+        });
+    }
+
+    public static SelectionSet mergeFilteredSelection(Field node, HashMap<String, Set<Field>> selectionSetTree) {
+        Field field;
+        SelectionSet.Builder selectionSetBuilder = SelectionSet.newSelectionSet();
+        SelectionSet.Builder childrenSelectionSetBuilder = SelectionSet.newSelectionSet();
+        if (ObjectUtils.isEmpty(selectionSetTree.get(node.getName()))) { // check if leaf node
+            field = new Field(node.getName(), node.getArguments());
+            selectionSetBuilder.selection(field);
+            return selectionSetBuilder.build();
+        }
+        List<SelectionSet> childrenSelectionSets = new ArrayList<>();
+        for (Field f : selectionSetTree.get(node.getName())) {
+            childrenSelectionSets.add(mergeFilteredSelection(f, selectionSetTree));
+        }
+        childrenSelectionSets
+                .stream()
+                .map(s -> s.getSelections())
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList())
+                .stream()
+                .forEach(childrenSelectionSetBuilder::selection);
+        field = new Field(node.getName(), node.getArguments(), childrenSelectionSetBuilder.build());
+        selectionSetBuilder.selection(field);
+
+        return selectionSetBuilder.build();
+    }
+
+    public SelectionSet getTransformedSelectionSet() {
+        if (operationType.name().equalsIgnoreCase("QUERY")
+                && canOptimizeQuery(filteredSelection)) {
+            SelectionSet.Builder mergedSelectionSetBuilder = SelectionSet.newSelectionSet();
+
+            HashMap<String, Set<Field>> selectionSetTree = new HashMap<>();
+            QueryOptimizer.createSelectionSetTree(filteredSelection, selectionSetTree);
+
+            filteredSelection.getSelections().stream()
+                    .map(rootNode -> (Field) rootNode)
+                    .filter(distinctByFieldName(Field::getName))
+                    .forEach(rootNode -> QueryOptimizer.mergeFilteredSelection(rootNode, selectionSetTree)
+                            .getSelections()
+                            .stream()
+                            .forEach(mergedSelectionSetBuilder::selection));
+            return mergedSelectionSetBuilder.build();
+        }
+        return filteredSelection;
+    }
+
+    private boolean canOptimizeQuery(SelectionSet selectionSet) {
+        if (serviceMetadata.hasFieldResolverDirective()) return false;
```

Comment:
What does this method do and why do we need to check this ? This problem will occur regardless of resolverDirective and whether the field will have directives or not.

Reply:
I can get rid of serviceMetadata.hasFieldResolverDirective() condition
But since I need to modify selectionSets, I need to create new Field objects. And on creating new Fields, there's no constructor provided that takes directives as an argument and so I was unable to return the correct query

Reply:
Two options
* There should be a builder ` Field.newField()`
* There should be transform method that takes a builder, this is prferable because this way you can only pass things you need to change. in this case, I think its selectionSet